### ISSUE TITLE
Allow BeforeAll Should Have the Option To Be Called Just Once in a Describe in FullyParallel Mode

### ISSUE THREAD

**Issue by awhisler** [microsoft/playwright#28814]
I really don't understand the purpose of the beforeAll in fullyParallel mode if it acts the exact same as beforeEach. If I understand correctly, both beforeAll and beforeEach will get called before EACH test if you are in full parallel mode. Please let me know if I am misunderstanding.

My request is to give the option to run beforeAll ONCE in a file even if full parallel mode. If I want things to be 'independent', I will use the beforeEach but I am finding myself constantly having to add `test.describe.configure({ mode: 'serial' })` because I want the beforeAll to run just once and it will not. Fullyparallel mode is much faster and it works for all my use cases except when I want to setup a scenario with a beforeAll but only once for the file. 

Example: How I use beforeAll is to create **one** new record. Then, in each test (let's say there are 5 tests) I do something different to it (I dont actually submit to _change_ the record but I interact with the record differently to create different test cases)

Example of Ideal:
'''
let botName;
test.describe('@Config : @Bots', () => {
    test.beforeAll(async ({ page }) => { // I want to create a bot
        await LoginPage.login({ page }, common.emailSuperAdmin);
        await MenuItem.clickConfigSubMenu({ page }, 'bots');
        botName = `AutomationBot${Math.random().toString().substring(2, 5)}`;
        await Bots.createAndPublishBot({ page }, botName, 'Personal Lines', 'Apartments');
      });
    test('validate new bot can be selected in report in UI', async ({ page }) => {
        await MenuItem.clickReports({ page });
        expect(await Reports.getBotNames({ page })).toContain(botName);
    });
    test('validate new bot can be inputted in report in API', async ({ page }) => {
       const status = await API.createRecord({page}, botName);
       expect(status).toEqual(200);
    });
});
'''

**Comment by awhisler** [comment]
If there is ANY other way to run some part of the code once in full parallel mode, I'm all ears!

**Comment by pavelfeldman** [comment]
> I really don't understand the purpose of the beforeAll in fullyParallel mode if it acts the exact same as beforeEach. If I understand correctly, both beforeAll and beforeEach will get called before EACH test if you are in full parallel mode. Please let me know if I am misunderstanding.

That's not the case, we will minimize the number of times we run beforeAll, you'll only get one per worker

It sounds like you are looking for a setup rather than before all, https://playwright.dev/docs/test-global-setup-teardown#setup. Setup will run once and will populate your databases.

**Comment by awhisler** [comment]
@pavelfeldman Thanks for the response! I need a setup PER spec file bc I need to know the details of that API call to do assertions in the test, so that will not work for my use case. But additionally, I was not able to get that to work for APIs.

**Comment by mxschmitt** [comment]
See https://github.com/microsoft/playwright/issues/28814#issuecomment-1872606467, having a project setup dependency for each test file would work for that.

It's expected that beforeAll can run multiple times if you have fullyParallel: true. What does not work for APIs in your case? Are you able to provide a repro or share more about your use-case?

**Comment by awhisler** [comment]
@mxschmitt How do I setup a project dependency for each file? [This](https://playwright.dev/docs/test-global-setup-teardown#setup) link seems like it's just one global file.

I tried to setup just a global one:
![image](https://github.com/microsoft/playwright/assets/2396250/3eca99bf-dc74-43fd-89e8-71a16a74a06e)
![image](https://github.com/microsoft/playwright/assets/2396250/73b00cd1-cad4-4e1a-b6d7-fb6a8e65157b)

And it never runs

**Comment by awhisler** [comment]
Here is a minimum project. I removed the API call in the global setup for confidentially but added a console log to clearly show it isn't working:
https://github.com/awhisler/example-for-debug

Run call: `URL=https://staging.dashboard.flyreel.co npx playwright test`

**Comment by mxschmitt** [comment]
You could create this multiple times for the files you want to have a setup project for: https://github.com/awhisler/example-for-debug/blob/a5c822b64342b699700e9adccbde2652db064f97/playwright.config.ts#L27-L32 with adding a testMatch on the specific test-file.

You need to move the `global.setup.ts` file into the `tests` directory, since your testsDir is `tests`.

**Comment by awhisler** [comment]
@mxschmitt got it, that worked! Thank you.

As for multiple files use case. How do I 'assign' it to a file? So I have the new file for a specific spec but I wouldn't want to add it to the dependencies of chrome because that would run for all files?
![image](https://github.com/microsoft/playwright/assets/2396250/33be4d4f-7175-4e55-8833-703827354d30)

**Comment by awhisler** [comment]
Weird, the global one worked but now I am getting this which I think was the API issue with global setup I was referencing prior
![image](https://github.com/microsoft/playwright/assets/2396250/713db76b-08ce-4c23-9b97-cf67f94ff43b)

**Comment by mxschmitt** [comment]
> How do I 'assign' it to a file? 

You would have to filter via [testMatch](https://playwright.dev/docs/api/class-testproject#test-project-test-match) in your chrome project. Then setup will run, setup for create-report.spec.ts and then create-report.spec.ts itself will run.

We usually would recommend now to rethink the test structure, since the test depend a lot on external factors.

**Comment by awhisler** [comment]
@mxschmitt it's not depending on other tests. It merely API setup for test data. I need to POST X thing before I can assert something within the actual test

**Comment by pavelfeldman** [comment]
To make sure I understand the problem:
- You have a bunch of test files
- One of them is called export-report.spec.ts and it requires certain setup
- You'd like to run all the tests in export-report.spec.ts (and in other files) in parallel

If that is the case, then the simplest way of achieving it is to have a single global setup, as I recommended originally. Think about it this way: fully parallel mode will run the tests in the order you can't control. Imagine that it runs first test in export-report.spec.ts among the first tests it runs and it runs the last test in export-report.spec.ts among the last tests it runs. I.e. it needs setup for this file to be in place at all times and there is nothing you can do about it. While it is running tests in export-report.spec.ts in parallel, it also runs tests from other files in parallel.

If that does not work for you, you can try making beforeAll idempotent. That way your setup routine implemented in beforeAll will run once and lazily behind your check. You can use file system to achieve idempotent behavior, but be careful with the atomic file operations.

**Comment by awhisler** [comment]
@pavelfeldman No, that is not my use case. Each test case has their own setup that is very unique to the test. BeforeAll doesn't work for me because I do not need to muddy the DB will creating a new 'thing' for EACH test cases (in parallel mode) in the spec file when I only need to create the 'thing' once and then validate different things on it in each test case. 

As for making the beforeAll idempotent, it is an API call of creating some object. How would I use a file system to achieve this?

**Comment by pavelfeldman** [comment]
> Each test case has their own setup that is very unique to the test. 

I'm confused why you want to use beforeAll in this case, beforeAll is shared across the test cases.

> As for making the beforeAll idempotent, it is an API call of creating some object. How would I use a file system to achieve this?

You would create a marker file once you've done something once.

**Comment by awhisler** [comment]
Well I only 'want' to use it here because thats what I've known in webdriverio and protractor back in the day as a way to setup the test.

**Comment by awhisler** [comment]
@pavelfeldman But Im having a problem using beforeAll for my use case because if I do, I need to use serial mode but browserstack does not put each test as a session if I am in serial mode. So I am blocked both ways with playwright's functionality. Any suggestions there?

![image](https://github.com/microsoft/playwright/assets/2396250/c1553160-74b3-4926-982d-62605a6cba41)

**Comment by pavelfeldman** [comment]
@awhisler I must admit I am lost with respect to your use case.

> No, that is not my use case. Each test case has their own setup that is very unique to the test.

Do you mean test file here? You seem to be using test case and test interchangeably, but I assume you are actually talking about the test file?

> The typical BeforeAll doesn't work for me because I do not need to muddy the DB will creating a new 'thing' for EACH test cases (in parallel mode) in the spec file when I only need to create the 'thing' once and then validate different things on it in each test case.

So you create (write) some data into the database once, and then you want to run a number of checks in parallel that would be querying (reading) this data and validating it. This assumes concurrent access to the same data, which we don't encourage. I would recommend to use default parallelism for this, create database entry in beforeAll and clean it up in afterAll, in each test file. If you have enough files, the performance win of the fully parallel will be negligible. Worst case you can balance them by hand. But this way at least, files run one after another, so you can create entry once in beforeAll and delete it in afterAll. Going fully parallel will make your database have all the entries for all the files created at the same time, which does not sound like something you want.

**Comment by awhisler** [comment]
@pavelfeldman Fair. Yes, I do mean test file. Each test file has unique setup.

Example of how I use this. I create 3 new things with a POST and on each test, I am validating different things in the search (search first name, search policy id, search partial name and make sure it searches correctly) **but**, if I was to use full parallel mode here with it creating a POST each time, then my partial searches will return more than I want to assert against. Yes, I could use "greaterThanOrEqualTo" but I don't love that and I would have to do loose assertions bc I wouldn't know if the new test started and created another thing that could match my current assertion.

![image](https://github.com/microsoft/playwright/assets/2396250/649a8a5c-1ef4-46c3-9a2f-1c4f6c9c98c0)

The thing that sucks with not using parallel mode is that browserstack will just have one giant session instead of each test or file as a session but I guess I need to sort that out separately.

**Comment by pavelfeldman** [comment]
Since you seem to be managing parallelism and shared data, here is another approach that demonstrates how data can be isolated by parallel workers. It might help with your use case. It is talking about authentication, but it is not much different from seeding the database with the data in principle.

https://playwright.dev/docs/auth#moderate-one-account-per-parallel-worker

> The thing that sucks with not using parallel mode is that browserstack will just have one giant session instead of each test or file as a session but I guess I need to sort that out separately.

As a fall back, Playwright will still do parallel, even if you don't do fully parallel. So with the browserstack, you should get as many parallel sessions as you have files. Assuming you have more than 10 files, it should already be giving you a fairly reasonable parallelism.

**Comment by awhisler** [comment]
I'll take a look!

I have 128 tests and 15 test files and in browserstack it shows at one giant session. Any thoughts?

**Comment by dgozman** [comment]
Closing as discussion seems resolved and there is no action item for Playwright here. If you still encounter problems, please file a new issue with a repro and link to this one.

**Comment by vitalets** [comment]
I was struggling with the same problem, and ended up with universal implementation of idempotent `beforeAll` hook [suggested](https://github.com/microsoft/playwright/issues/28814#issuecomment-1880335574) by @pavelfeldman.

The goal of @awhisler's code to create a bot once for all tests could be achieved like this:

```ts
import { globalCache } from '@vitalets/global-cache'; // <-- import globalCache

let botName;
test.describe('@Config : @Bots', () => {
    test.beforeAll(async ({ browser }) => {
       botName = await globalCache.get('bot-name', async () => {  // <-- Create a bot ONCE and cache for all tests
          const page = await browser.newPage();
          await LoginPage.login({ page }, common.emailSuperAdmin);
          await MenuItem.clickConfigSubMenu({ page }, 'bots');
          const botName = `AutomationBot${Math.random().toString().substring(2, 5)}`;
          await Bots.createAndPublishBot({ page }, botName, 'Personal Lines', 'Apartments');
          return botName;
       });
    });

    test('validate new bot can be selected in report in UI', async ({ page }) => {
        await MenuItem.clickReports({ page });
        expect(await Reports.getBotNames({ page })).toContain(botName);
    });

    test('validate new bot can be inputted in report in API', async ({ page }) => {
       const status = await API.createRecord({page}, botName);
       expect(status).toEqual(200);
    });
});
```

Here're more details: https://github.com/microsoft/playwright/issues/22520#issuecomment-3167474277

**Comment by jeannelmah** [comment]
Facing this issue as well. beforeAll and beforeEach within a spec have no difference when running in parallel since both of them are run before each test in a spec. It doesn't make sense. Shouldn't it be the case where the beforeAll runs once before the rest of the spec, then beforeEach runs before each test? Like sometimes you don't want to run this spec and thus have no need to setup data globally as you're only interested in this data for the specific spec.
